Allow redefinition of where logging goes

thirstyice opened the pull request:
Sometimes I want to log to places other than `Serial` (when I have a display connected, or when my adapter is attached to `Serial1` for example)

Also adds guards so things don't break when the user disables `Serial`

hideakitai commented on Artnet/Receiver.h:
```diff
@@ -293,12 +293,22 @@ class Receiver_
         this->stream = &s;
     }
 
+#if defined(NO_GLOBAL_INSTANCES) || defined(NO_GLOBAL_SERIAL)
```
Could you let me know what these preprocessors are?

thirstyice replied:
They're the same macros that allow disabling `Serial`et al in the esp32 arduino core (see [here](https://github.com/arduino/arduino-esp32/blob/21b88659b9ded3fcc1082f23f498bc2a04cd4f1b/cores/esp32/HardwareSerial.cpp#L98C1-L98C64)).
I had assumed when I was developing this that they were standard across all arduino cores, but it seems that they are not.

hideakitai commented on Artnet/Receiver.h:
```diff
@@ -293,12 +293,22 @@ class Receiver_
     }
 
+#if defined(NO_GLOBAL_INSTANCES) || defined(NO_GLOBAL_SERIAL)
+    Print* log = nullptr; // No Serial -> no output
```
I think if the user doesn’t set this, the program crashes. (nullptr->print() will be called)

thirstyice replied:
Upon further research: this is actually undefined behaviour (not neccesarily a crash), which is why I didn't catch it earlier.
Solution coming soon.

hideakitai reviewed (commented):
Thank you, please check the comments.

thirstyice commented:
Added a subclass of `Print` that does nothing, which is now the default log output. The user can then choose if + where to log by calling `logOutputTo`.

Removed `b_verbose`, as it seemed redundant to have to call 2 separate functions to enable logging.

This way we also don't have to worry about what happens if `Serial` is undefined, that becomes the user's problem.

hideakitai reviewed (approved):
Almost LGTM. Please check the comment

hideakitai commented on Artnet/Receiver.h:
```diff
@@ -299,6 +284,8 @@ class Receiver_
         this->stream = &s;
     }
 
+    Print* log = &NoLog;
```
Oh well, the default was verbose = false, so it is ok if the default is NoSerial—my mistake.

hideakitai commented on Artnet/Common.h:
```diff
@@ -85,6 +85,13 @@ struct Destination
     uint8_t universe;
 };
 
+class NoPrint : public Print {
+    size_t write(uint8_t) {
+        return 0;
+    }
+};
+NoPrint NoLog;
```
Also, I think it's better to make `NoLog` to `static NoPrint no_log;` (`static` is for the size optimization for multiple receivers, and renaming to snake case is in terms of variable naming uniformity)
